src/core/geometry/qgssfcgalgeometry.h

```diff
@@ -0,0 +1,516 @@
+     */
+    bool addZValue( double zValue = 0, QString *errorMsg = nullptr );
+
+    /**
+     * Adds a m-dimension to the geometry, initialized to a preset value (existing M values remains unchanged).
+     * \return true if success
+     * \param mValue m value to use
+     * \param errorMsg pointer to QString to receive the error message if any
+     */
+    bool addMValue( double mValue = 0, QString *errorMsg = nullptr );
+
+    /**
+     * Drops the z coordinate of the geometry
+     * \return true if success
+     * \param errorMsg pointer to QString to receive the error message if any
+     */
+    bool dropZValue( QString *errorMsg = nullptr );
+
+    /**
+     * Drops the m coordinate of the geometry
+     * \return true if success
+     * \param errorMsg pointer to QString to receive the error message if any
+     */
+    bool dropMValue( QString *errorMsg = nullptr );
+
+    /**
+     * Swaps the x and y coordinates of the geometry
+     * \param errorMsg pointer to QString to receive the error message if any
+     */
+    void swapXy( QString *errorMsg = nullptr );
+
+    /**
+     * Checks if \a geom is valid.
+     *
+     * If the geometry is invalid, \a errorMsg will be filled with the reported geometry error.
+     *
+     * \param errorMsg Error message returned by SFGCAL
+     * \param flags with Qgis::GeometryValidityFlag::AllowSelfTouchingHoles, specifies whether self-touching holes are permitted.
+     *        OGC validity states that self-touching holes are NOT permitted, whilst other
+     *        vendor validity checks (e.g. ESRI) permit self-touching holes.
+     */
+    bool isValid( Qgis::GeometryValidityFlags flags, QString *errorMsg = nullptr ) const;
+
+    /**
+     * Checks if \a geom is empty.
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    bool isEmpty( QString *errorMsg = nullptr ) const;
+
+    /**
+     * Computes the area of \a geom.
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    double area( QString *errorMsg = nullptr ) const;
+
+    /**
+     * Computes the max length of \a geom.
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    double length( QString *errorMsg = nullptr ) const;
+
+    /**
+     * Checks this geometry is simple.
+     *
+     * \see QgsSfcgalEngine::isSimple
+     *
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    bool isSimple( QString *errorMsg = nullptr ) const;
+
+    /**
+     * Calculates the centroid of this geometry.
+
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    QgsPoint centroid( QString *errorMsg ) const;
+
+    /**
+     * Translate this geometry by vector \a translation.
+     *
+     * \param translation translation vector (2D or 3D)
+     * \param errorMsg Error message returned by SFGCAL
+     */
+    QgsSfcgalGeometry *translate( const QgsPoint &translation, QString *errorMsg = nullptr ) const;
```

Comment:
```suggestion
    QgsSfcgalGeometry *translate( const QgsVector3D &translation, QString *errorMsg = nullptr ) const;
```

Reply:
Ok but why?

Reply:
because you translate by a vector, not a point.

Reply:
ok